presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java

```diff
@@ -0,0 +1,236 @@
+import java.util.Optional;
+
+import static com.facebook.presto.metadata.Signature.comparableTypeParameter;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.type.ArrayType.toStackRepresentation;
+import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.slice.Slices.utf8Slice;
+import static java.lang.String.format;
+
+public final class ArrayDistinctFunction
+        extends ParametricScalar
+{
+    public static final ArrayDistinctFunction ARRAY_DISTINCT_FUNCTION = new ArrayDistinctFunction();
+    private static final String FUNCTION_NAME = "array_distinct";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(comparableTypeParameter("E")), "array<E>", ImmutableList.of("array<E>"), false, false);
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayDistinctFunction.class, "distinct", Type.class, Slice.class);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get();
+    private static final CollectionType COLLECTION_TYPE = OBJECT_MAPPER.getTypeFactory().constructCollectionType(List.class, Object.class);
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Remove duplicate values from the given array";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        checkArgument(types.size() == 1, format("%s expects only one argument", FUNCTION_NAME));
+        Type type = types.get("E");
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(type);
+        Signature signature = new Signature(FUNCTION_NAME,
+                parameterizedTypeName("array", type.getTypeSignature()),
+                parameterizedTypeName("array", type.getTypeSignature()));
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle, isDeterministic(), false, ImmutableList.of(false));
+    }
+
+    private static Block listToBlock(final Type type, List<Object> elements)
+    {
+        BlockBuilder elementsBlockBuilder;
+        if (type instanceof AbstractFixedWidthType) {
+            elementsBlockBuilder = ((AbstractFixedWidthType) type).createFixedSizeBlockBuilder(elements.size());
+        }
+        else {
+            elementsBlockBuilder = type.createBlockBuilder(new BlockBuilderStatus());
+        }
+
+        Class<?> javaType = type.getJavaType();
+        if (javaType == boolean.class) {
```

Comment:
I think this will be less code, if you switch the order of the loops and ifs, so that it's:
```
for (Object e : elements) {
if (e == null) {
appendNull()
continue;
}
if (type == ....)
}
```